### PR TITLE
add skeletons for MantleBackup data synchronization

### PR DESCRIPTION
**Pull request by ushitora-anqou**

### CODE REVIEW

**Review by satoru-takeuchi — CHANGES_REQUESTED** [review]

**Comment by ushitora-anqou** [comment]
Note: I need to change the commit messages of some commits to reflect their fixed contents.

File: internal/controller/mantlebackup_controller.go [hunk 1]
```diff
@@ -344,26 +347,28 @@ func (r *MantleBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
+		return ctrl.Result{}, nil
+	}
+
 	if r.role == RolePrimary {
-		return r.replicate(ctx, logger, &backup)
+		result, err := r.replicateManifests(ctx, logger, &backup)
+		if err != nil || result != (ctrl.Result{}) {
+			return result, err
+		}
+		prepare, result, err := r.prepareForDataSynchronization(ctx, &backup, r.primarySettings.Client)
```
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller.go:364]: I couldn't come up with anything good. I'll use `prepareResult`.

File: internal/controller/mantlebackup_controller.go [hunk 2]
```diff
@@ -292,6 +293,17 @@ func (r *MantleBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, err
 	}
 
+	// Make sure that the reconciled MantleBackup is created not by the secondary mantle but by the primary mantle.
```
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller.go]: fixed in https://github.com/cybozu-go/mantle/pull/46/commits/c501f70ca1d62f3d89fc54834bd31d1d3accadf7

File: internal/controller/mantlebackup_controller.go [hunk 3]
```diff
@@ -543,3 +539,96 @@ func (r *MantleBackupReconciler) provisionRBDSnapshot(
 
 	return nil
 }
+
+func validateMantleBackupAsPrimary(backup *mantlev1.MantleBackup) error {
```
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller.go]: fixed in https://github.com/cybozu-go/mantle/pull/46/commits/c501f70ca1d62f3d89fc54834bd31d1d3accadf7

File: internal/controller/mantlebackup_controller.go [hunk 4]
```diff
@@ -543,3 +539,96 @@ func (r *MantleBackupReconciler) provisionRBDSnapshot(
...
+			return fmt.Errorf("both %s and %s labels are found", labelLocalBackupTargetPVCUID, labelRemoteBackupTargetPVCUID)
+		}
+	}
+	if _, ok := backup.Annotations[annotRemoteUID]; ok {
+		return fmt.Errorf("annotation %s found", annotRemoteUID)
+	}
+	return nil
+}
+
+func (r *MantleBackupReconciler) removeFinalizerAndRBDSnapshot(
```
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller.go]: fixed in https://github.com/cybozu-go/mantle/pull/46/commits/00cae6d19e7242d59c82ce4e1e371612ed6faed7

File: internal/controller/mantlebackup_controller.go [hunk 5]
```diff
@@ -307,23 +319,14 @@ func (r *MantleBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
-				}
-			}
-
-			controllerutil.RemoveFinalizer(&backup, MantleBackupFinalizerName)
-			err = r.Update(ctx, &backup)
-			if err != nil {
-				logger.Error("failed to remove finalizer", "finalizer", MantleBackupFinalizerName, "error", err)
-				return ctrl.Result{}, err
-			}
+		if _, ok := backup.GetAnnotations()[annotDiffTo]; ok {
```
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller.go]: fixed in https://github.com/cybozu-go/mantle/pull/46/commits/851b98f0bad03cbf65c7df4de0492dbe563d124d and https://github.com/cybozu-go/mantle/pull/46/commits/b9e49ce3002e33471633239e9481d216de905b9a

File: internal/controller/mantlebackup_controller.go [hunk 6]
```diff
@@ -344,26 +347,28 @@ func (r *MantleBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
+			return result, err
+		}
+		prepare, result, err := r.prepareForDataSynchronization(ctx, &backup, r.primarySettings.Client)
+		if err != nil || result != (ctrl.Result{}) {
+			return result, err
+		}
+		if prepare.isSecondaryMantleBackupReadyToUse {
+			return r.primaryCleanup(ctx, logger, &backup)
+		}
+		return r.export(ctx, &backup, r.primarySettings.Client, prepare)
```
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller.go]: fixed in https://github.com/cybozu-go/mantle/pull/46/commits/8a87df319ee6eca9420ac5efd359cc6f7d3ec3b7

File: internal/controller/mantlebackup_controller.go [hunk 7]
```diff
@@ -543,3 +539,96 @@ func (r *MantleBackupReconciler) provisionRBDSnapshot(
...
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+type dataSyncPrepareResult struct {
+	isIncremental                     bool
+	isSecondaryMantleBackupReadyToUse bool
+	diffFrom                          *mantlev1.MantleBackup // non-nil value iff isFull is false.
```
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller.go]: fixed in https://github.com/cybozu-go/mantle/pull/46/commits/ae497fdcb49dda27149ab817956a0f839e688ce0

File: internal/controller/mantlebackup_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller.go]: fixed in https://github.com/cybozu-go/mantle/pull/46/commits/21c9c54af018f0742760731e688f0dd3ab69c732

**Comment by ushitora-anqou** [comment]
@satoru-takeuchi I think I've fixed all the issues you mentioned. Could you review this PR again?

Note that I need to squash my fixup commits, so please don't merge it yet, even if the code looks good to you.

**Comment by ushitora-anqou** [comment]
@satoru-takeuchi I think I've fixed all the issues you mentioned. Could you review this PR again?

Note that I need to squash my fixup commits, so please don't merge it yet, even if the code looks good to you.

**Review by satoru-takeuchi — APPROVED** [review]
LGTM. Please squash commits.

**Comment by ushitora-anqou** [comment]
Note: comparison: https://github.com/cybozu-go/mantle/compare/2e874d56d3467a52149d3755b4b2d2b3fe33c910..416bad5ccca987c47dfbca4148f62ebf1db2d598

**Pull request merged**